# Hygeia Case-Control Analysis

>TODO - ADD DESCRIPTION OF WHAT Hygeia is.

## Single Group analysis

> TODO - ADD description of Single Group analysis

## Case-Control Analysis

>TODO - ADD description of Case-control analysis

## Running Hygeia

### Installation

* Python 3.8

```bash
pip install -r requirements.txt
```

### Running Hygeia

#### Data-Preprocessing

The following script (in DataProcessing) generates for each chromosome - a text file containing (i) CpG positions, (ii) methylated reads control group, (iii) methylated reads case group, (iv) total reads control group, and (v) total reads case group.
These are the input count files for Hygeia. In this script, we use all CpG sites from the reference genome.
We use the methylation data generated from gemBS.
However, Hygeia should work as long as the generated files are indexed at the same CpG positions (possibly with zero methylated or total reads).

To run the below command without command line arguments, unzip the input file (containing gemBS data of the newborn and adult, parameter estimates of the single group model for selected chromosomes and all CpG sites from the reference genome) and save it under ./test_data

```bash
bin/hygeia preprocess \
  -cpg_file_path ./tmp/test_data/cpg.tsv.gz \
  -case_data_path ./tmp/test_data/EGAZ00001016574_90_cpg.txt.gz \
  -case_id_names EGAZ00001016574_90 \
  -control_data_path ./tmp/test_data/EGAZ00001016575_90_cpg.txt.gz \
  -control_id_names EGAZ00001016575_90 \
  -output_path ./results/preprocessed
```

Here are further details on each argument:

```bash
-cpg_file_path: path to file that contains all CpG sites that one wants to use (will be imputed if missing in methylation count files). File should be tab separated.
-output_path: directory where the count files are saved
-case_data_path: list of paths of the methylation data for the case group as generated by gemBS.
-case_id_names: names of the case ids that are used in the gemBS output files
-control_data_path: list of paths of the methylation data for the control group as generated by gemBS.
-control_id_names: names of the control ids that are used in the gemBS output files
```

#### Case-Control Model Inference

The following script (in CaseControlScripts) generates for a given chromosome the full Hygeia outputs.
This includes the particles (duration and METEOR/regime) for the backward simulation.
The results can be seen as one Monte Carlo sample (specified by the seed flag) of the algorithm.
These outputs can be used to test/discover for DMPs/DMRs with further scripts.

```bash
bin/hygeia infer \
  -results_dir ./results/out \
  -data_dir ./results/preprocessed \
  -single_group_dir ../SingleGroup/result/params \
  -chrom 22
```

Here are further details on each argument

```bash
-results_dir: path dir where to save the results
-data_dir: path where the count data is saved (e.g. as outputs from preprocess_data.py)
-single_group_dir: path where the model parameters of the single group model are saved 
-chrom: chromosome number for the analysis
```

There are also a couple of model specific or inference specific flags that can be adjusted.



#### Case-Control Model Inference on simulated data

To simulate data from a case-control model and test the inference algorithm, run the following command 

```bash
python hygeia/TwoGroup/simulate_two_groups.py
```
Useful command line arguments are

```bash
-regimes_config: configuration determining the shape of the METEORs (e.g. 8)
-num_observations: number of CpG sites to simulate
-model_dir: path dir where to save the results
-num_resampled_particles: number of resampled particles with higher number (e.g. 50) yielding better accuracy at increased computational cost compared to a lower number (e.g. 20) t
-num_samples_backward: number of particles in the backward pass (e.g. 25)
```

